File: eco-core/core-plugin/src/main/java/com/willfp/ecoarmor/upgrades/Tiers.java
package com.willfp.ecoarmor.upgrades;

import com.google.common.collect.BiMap;
import com.google.common.collect.HashBiMap;
import com.google.common.collect.ImmutableList;
import com.willfp.eco.core.config.interfaces.Config;
import com.willfp.eco.core.config.updating.ConfigUpdater;
import com.willfp.ecoarmor.EcoArmorPlugin;
import lombok.Getter;
import lombok.experimental.UtilityClass;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;

@UtilityClass
public class Tiers {
    /**
     * Registered tiers.
     */
    private static final BiMap<String, Tier> BY_ID = HashBiMap.create();

    /**
     * Default tier.
     */
    @Getter
    private static Tier defaultTier;

    /**
     * Get {@link Tiers} matching ID.
     *
     * @param name The name to search for.
     * @return The matching {@link Tiers}, or null if not found.
     */
    @Nullable
    public static Tier getByID(@Nullable final String name) {
        return BY_ID.get(name);
    }

    /**
     * Get all registered {@link Tiers}s.
     *
     * @return A list of all {@link Tiers}s.
     */
    public static List<Tier> values() {
        return ImmutableList.copyOf(BY_ID.values());
    }

    /**
     * Add new {@link Tier} to EcoArmor.
     *
     * @param tier The {@link Tier} to add.
     */
    public static void addNewTier(@NotNull final Tier tier) {
        BY_ID.remove(tier.getId());
        BY_ID.put(tier.getId(), tier);
    }

    /**
     * Update.
     *
     * @param plugin Instance of EcoArmor.
     */
    @ConfigUpdater
    public static void reload(@NotNull final EcoArmorPlugin plugin) {
        BY_ID.clear();

        for (Config tierConfig : plugin.getEcoArmorYml().getSubsections("tiers")) {
            new Tier(tierConfig, plugin);
        }

        defaultTier = Tiers.getByID("default");
    }

    static {
        reload(EcoArmorPlugin.getInstance());
    }
}

File: eco-core/core-plugin/src/main/kotlin/com/willfp/ecoarmor/upgrades/Tiers.kt
package com.willfp.ecoarmor.upgrades

import com.google.common.collect.ImmutableList
import com.willfp.eco.core.config.interfaces.Config
import com.willfp.eco.core.registry.Registry
import com.willfp.libreforge.loader.LibreforgePlugin
import com.willfp.libreforge.loader.configs.ConfigCategory
import com.willfp.libreforge.loader.configs.LegacyLocation

object Tiers : ConfigCategory("tier", "tiers") {
    /**
     * Registered tiers.
     */
    private val registry = Registry<Tier>()

    private var cachedValues: List<Tier> = emptyList()

    override val supportsSharing = false

    override val legacyLocation = LegacyLocation(
        "ecoarmor.yml",
        "tiers"
    )

    /**
     * Default tier.
     */
    @JvmStatic
    lateinit var defaultTier: Tier

    /**
     * Get [Tiers] matching ID.
     *
     * @param name The name to search for.
     * @return The matching [Tiers], or null if not found.
     */
    @JvmStatic
    fun getByID(name: String?): Tier? {
        if (name == null) {
            return null
        }

        return registry[name]
    }

    /**
     * Get all registered [Tiers]s.
     *
     * @return A list of all [Tiers]s.
     */
    @JvmStatic
    fun values(): List<Tier> {
        return cachedValues
    }

    override fun clear(plugin: LibreforgePlugin) {
        registry.clear()
        cachedValues = emptyList()
    }

    override fun acceptConfig(plugin: LibreforgePlugin, id: String, config: Config) {
        registry.register(Tier(id, config))
    }

    override fun afterReload(plugin: LibreforgePlugin) {
        cachedValues = ImmutableList.copyOf(registry.values())
        defaultTier = getByID("default")!!
    }
}
